packages/deployments/contracts/contracts/messaging/RootManager.sol

```diff
@@ -197,23 +197,31 @@ contract RootManager is ProposedOwnable, IRootManager, WatcherClient, DomainInde
     // Dequeue verified roots from the queue and insert into the tree.
     (bytes32 _aggregateRoot, uint256 _count) = dequeue();
 
-    // Sanity check: make sure we are not propagating a redundant aggregate root.
-    require(_aggregateRoot != lastPropagatedRoot, "redundant root");
-    lastPropagatedRoot = _aggregateRoot;
-
     uint256 sum = msg.value;
     for (uint32 i; i < _numDomains; ) {
-      // Try to send the message with appropriate encoded data and fees
-      // Continue on revert, but emit an event
-      try
-        IHubConnector(_connectors[i]).sendMessage{value: _fees[i]}(abi.encodePacked(_aggregateRoot), _encodedData[i])
-      {
-        // NOTE: This will ensure there is sufficient msg.value for all fees before calling `sendMessage`
-        // This will revert as soon as there are insufficient fees for call i, even if call n > i has
-        // sufficient budget, this function will revert
-        sum -= _fees[i];
-      } catch {
+      // Sanity check: make sure we are not propagating a redundant aggregate root.
+      // If the last propagate root is the aggregate, treat as a failure.
+      if (_aggregateRoot == lastPropagatedRoot[domains[i]]) {
+        // Duplicate -- emit fail
         emit PropagateFailed(domains[i], _connectors[i]);
+      } else {
+        // Roots are different, try to send via connector to spoke
+
+        // Try to send the message with appropriate encoded data and fees
+        // Continue on revert, but emit an event
+        try
+          IHubConnector(_connectors[i]).sendMessage{value: _fees[i]}(abi.encodePacked(_aggregateRoot), _encodedData[i])
+        {
+          // Update last sent root IFF successful
+          lastPropagatedRoot[domains[i]] = _aggregateRoot;
+
```

Comment:
[Optional]
the `RootPropagated` is now a bit misleading since it  the root was propagated on ALL domains.
would it make sense to move and rename `emit RootPropagated(_aggregateRoot, _count, domainsHash);` here?
to: `emit DomainRootPropagated(_aggregateRoot, _count, domainsHash, domains[i], _connectors[i]);`
feel free to remove unnecessary params from the `DomainRootPropagated` event

I'm not sure how would this impact individual L2/Sidechain relayers that trigger on this event.

Reply:
Emitting for each domain would add to the O(N) cost here.

Personally, I don't think it will mislead. If we fail to propagate to any 1 domain, then that 1 domain will never emit an `AggregateRootReceived` event:
https://github.com/connext/nxtp/blob/10e5df93b125b508bb53cb6393d6a836d5824391/packages/deployments/contracts/contracts/messaging/connectors/SpokeConnector.sol#L443

I would rather just treat this event as an indicator that "propagate was attempted for all domains"

And use the `PropagateFailed` event as an indicator that it failed for any particular domain. Subgraph mapping can use the failure event to track domains for which we failed to send root.

Reply:
Worth saying this is a good note for the edge case in which ALL domains failed - in which case, we'd be emitting `RootPropagated` and N `PropagateFailed` events which would be kinda redundant/pointless